src/lifecycles/UIComponentLifecycle.ts

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2023 Mikhail Aheichyk
+Copyright 2023 Nordeck IT + Consulting GmbH.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/**
+ * UI Component lifecycle events.
+ */
+export enum UIComponentLifecycle {
+    /**
+     * An event to check if component should be shown by the Element.
+     */
+    ShouldShowComponent = "should_show_component"
+}
+
+/**
+ * Opts object containing the result of whether the component should be shown or not.
+ */
+export type ShouldShowUIComponentOps = {
+    /**
+     * Should be true to show the component, false to hide it. If not defined, then the default value from
+     * the Element should be applied.
+     */
+    shouldShowComponent: boolean | undefined;
+};
+
+/**
+ * Listener for "should_show_component" request
+ */
+export type ShouldShowUIComponentListener = (
+    shouldShowUIComponentOps: ShouldShowUIComponentOps,
+    component: UIComponent,
+    matrixUserId: string
+) => void;
+
+export enum UIComponent {
```

Comment:
This seem to be a copy/paste of `UIFeature.ts` in the `matrix-react-sdk`. 
Before considering this i'd want to explore ways to keep the two in sync. I'm actually unsure what value this brings over the existing possible customisations?

Reply:
Yes, that is true, I have mentioned this in my first comment. Maybe a solution could be to change react-sdk to use a `UIComponent` from this PR? Or any other idea?

Based on this https://github.com/matrix-org/matrix-react-sdk/pull/9931#issuecomment-1387623808 we may think that component visibility customization API is deprecated in favor of module API.

This solution would simplify our deployment, maybe quite a lot. We plan to put all visibility customizations and landing for guests in a separate module.